### ISSUE TITLE
Changing order of trait bounds changes available impls

### ISSUE THREAD

**Issue by Pratyush** [rust-lang/rust#75691]
I tried this code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7036047834cc394106c643e5ca974e69)):

```rust
use core::ops::AddAssign;

trait TraitA<B>: Copy + AddAssign<B> + AddAssign<Self> {}

trait TraitC<B> { type Thing: TraitA<B>; }

fn use_c<B, C: TraitC<B>>(mut a: C::Thing, b: B) {
	a += b; // errors
	a += a;
}
```

I expected this to compile, because the trait bounds on `TraitA` are sufficient. Instead, `rustc` fails to find the `AddAssign<B>` bound for `C::Thing`, and errors out with the following error message:
```
  |
7 | fn use_c<B, C: TraitC<B>>(mut a: C::Thing, b: B) {
  |          - this type parameter
8 |     a += b; // errors
  |          ^ expected associated type, found type parameter `B`
  |
  = note: expected associated type `<C as TraitC<B>>::Thing`
              found type parameter `B`
help: consider further restricting this bound
  |
7 | fn use_c<B, C: TraitC<B> + TraitC<B, Thing = B>>(mut a: C::Thing, b: B) {
  |                          ^^^^^^^^^^^^^^^^^^^^^^
```

(Note that the `help` message, while technically correct, is actually unhelpful here.)

### Meta

This bug occurs in the latest stable, beta, and nightly (this can be verified on the playground link).

This might be related to https://github.com/rust-lang/rust/issues/41756.

**Comment by Pratyush** [comment]
Update: this compiles successfully with `chalk`; the output of
```bash
rustc +nightly --crate-name test_chalk --edition=2018 src/lib.rs --crate-type lib --emit=dep-info,metadata -C embed-bitcode=no -C debuginfo=2 -C metadata=1ff6265fd5bdefac -C extra-filename=-1ff6265fd5bdefac --out-dir /tmp/test_chalk/target/debug/deps -C incremental=/tmp/test_chalk/target/debug/incremental -L dependency=/tmp/test_chalk/target/debug/deps -C target-cpu=native -Z chalk=yes
```

is 

```
warning: function is never used: `use_c`
 --> src/lib.rs:7:4
  |
7 | fn use_c<B, C: TraitC<B>>(mut a: C::Thing, b: B) {
  |    ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

**Comment by Pratyush** [comment]
Update: this compiles on the latest nightly. Is there a way to figure out what changed?

**Comment by lzutao** [comment]
You could try out bisect-rustc to find which nightly and possibly which PR fix the problem: https://github.com/rust-lang/cargo-bisect-rustc/blob/master/TUTORIAL.md

**Comment by Pratyush** [comment]
Ahh I think it was this (see “ambiguous projection bounds”) https://github.com/rust-lang/rust/pull/73905#issuecomment-674438668

**Comment by Pratyush** [comment]
Can confirm that this is fixed due to #73905; closing.
